modules/cudf/src/dataframe/concat.ts

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2021, NVIDIA CORPORATION.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {DataFrame, SeriesMap} from '../data_frame';
+import {Series} from '../series';
+import {Table} from '../table';
+import {DataType} from '../types/dtypes';
+import {CommonTypes, findCommonType} from '../types/mappings';
+
+// clang-format off
+export type ConcatResult<D extends DataFrame, T extends unknown[]> =
+  T extends []                      ? D['types'] :
+  T extends [DataFrame]             ? CommonTypes<D['types'], T[0]['types']> :
+  T extends [DataFrame, ...infer U] ? CommonTypes<D['types'], ConcatResult<T[0], U>> :
+                                      D['types'] ;
+// clang-format on
+
+export function concat<TFirst extends DataFrame, TRest extends DataFrame[]>(first: TFirst,
+                                                                            ...others: TRest) {
+  const dfs              = [first, ...others] as [TFirst, ...TRest];
+  const all_column_names = [...dfs
+                              .reduce((namesMap, df) => df.names.reduce(
+                                        (namesMap, name) => namesMap.set(name, true), namesMap),
+                                      new Map<string, boolean>())
+                              .keys()];
```

Comment:
I think maybe this could be simplified?
```
const all_column_names = [...new Set(dfs.reduce((names, df) => names.concat(df.names), [])).keys()]
```

Reply:
```
No overload matches this call.
  Overload 1 of 2, '(...items: ConcatArray<never>[]): never[]', gave the following error.
    Argument of type 'readonly string[]' is not assignable to parameter of type 'ConcatArray<never>'.
      The types returned by 'slice(...)' are incompatible between these types.
        Type 'string[]' is not assignable to type 'never[]'.
          Type 'string' is not assignable to type 'never'.
  Overload 2 of 2, '(...items: ConcatArray<never>[]): never[]', gave the following error.
    Argument of type 'readonly string[]' is not assignable to parameter of type 'ConcatArray<never>'.ts(2769)
 ```
 
 Over `(df.names)`
 
 As parameter `names` ends up being `never[]`. Unless I'm misunderstanding, isn't `dfs` just an array of `DataFrame`?

Reply:
> isn't dfs just an array of DataFrame

I thought so. FWIW I was just basing my suggestion on some toy experiments:

https://www.typescriptlang.org/play?#code/MYewdgzgLgBAhjAvDA3mOBbAphAXDAbQCIAzEEIgGhiICM4AnIgXQF8AoUSWWpVdbHkJ1GVGvQBeLDp3DQYAExIQ+BONVrN2s7jACuYAJYBHPVgD6AnKoB0dsFgDuMAMpYoACiUQbDLAr1gLA8PKwhqJQBKJAA+GDCbLmA4TyUbMMjqAmZIyJsAaywATwgPSK0uCBAAGywbapAAcw8DEzNLTBxIoA

So it's possible (well, definitely) missed some detail

Reply:
Spoke with Bryan, found a stackoverflow thread that ran into this too, just define `names` as type `string[]`.